src/payment/strategies/cybersource/cybersource-threedsecure-payment-processor.ts

```diff
@@ -0,0 +1,234 @@
+                        .pipe(take(1))
+                        .subscribe((event: CardinalEventResponse) => {
+                            if (event.type.step === CardinalPaymentStep.SETUP) {
+                                if (!event.status) {
+                                    reject(new MissingDataError(MissingDataErrorType.MissingPaymentMethod));
+                                }
+
+                                this._isSetupCompleted = true;
+                                resolve();
+                            }
+                        });
+                });
+            }).then(() => this._store.getState());
+    }
+
+    execute(payment: OrderPaymentRequestBody, order: OrderRequestBody, paymentData: CreditCardInstrument, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        if (!this._Cardinal) {
+            return Promise.reject(new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized));
+        }
+
+        return this._Cardinal.trigger(CardinalTriggerEvents.BIN_PROCCESS, paymentData.ccNumber).then(result => {
+            if (result && result.Status) {
+                return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+                    .then(() =>
+                        this._store.dispatch(
+                            this._paymentActionCreator.submitPayment({ ...payment, paymentData })
+                        )
+                    ).catch(error => {
+                        if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'enrolled_card' })) {
+                            return Promise.reject(error);
+                        }
+
+                        if (!this._Cardinal) {
+                            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+                        }
+
+                        const continueObject = {
+                            AcsUrl: error.body.three_ds_result.acs_url,
+                            Payload: error.body.three_ds_result.merchant_data,
+                        };
+
+                        const partialOrder = {
+                            OrderDetails: {
+                                TransactionId: error.body.three_ds_result.payer_auth_request,
+                            },
+                        };
+
+                        this._Cardinal.continue(CardinalPaymentBrand.CCA, continueObject, partialOrder);
+
+                        // If credit card is enrolled in 3DS Cybersource will handle the rest of the flow
+                        return new Promise<string>((resolve, reject) => {
+                            this._cardinalEvent$
+                                .pipe(take(1))
+                                .subscribe((event: CardinalEventResponse) => {
+                                    if (event.type.step === CardinalPaymentStep.AUTHORIZATION) {
+                                        if (event.status) {
```

Comment:
Instead of checking the event.type here, can we add a RxJS fileter? Check this example: https://www.learnrxjs.io/operators/filtering/filter.html